File: Cpp_src/medium/longest_substring_without_repeating_characters_3.cc
#include <iostream>
#include <unordered_map>
#include <unordered_set>
using namespace std;

// Use slide window with hash set
// Time Complexity: O(2n) = O(n)
// Space Complexity:O(k)
// Runtime: 32 ms, faster than 33.86%
// Memory Usage: 13.3 MB, less than 30.85%
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_set<char> hash_set;
        unordered_set<char>::iterator iter;
        int result = 0;
        int i = 0, j = 0, length = s.size();
        while (i < length && j < length)
        {
            iter = hash_set.find(s.at(j));
            if (iter == hash_set.end())
            {
                hash_set.insert(s.at(j++));
                result = max(result, j - i);
            }
            else
            {
                hash_set.erase(s.at(i++));
            }
        }

        return result;
    }
};

// Use slide window (Assuming ASCII 128)
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 12 ms, faster than 83.19%
// Memory Usage: 9 MB, less than 98.51%
class Solution2
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int result = 0;
        int length = s.size();
        int index[128] = {0};
        for (int i = 0, j = 0; j < length; j++)
        {
            i = max(index[(int)s.at(j)], i);
            result = max(result, j - i + 1);
            index[(int)s.at(j)] = j + 1; // 1 for after same char position
        }
        return result;
    }
};

File: Golang_src/medium/longestSubstringWithoutRepeatingCharacters_3.go
package medium

// Use slide window:
// Time Complexity: O(n)
// Space Complexity:O(k)
// Runtime: 4 ms, faster than 90.23%
// Memory Usage: 2.6 MB, less than 100.00%
func lengthOfLongestSubstring(s string) int {
	if len(s) == 0 {
		return 0
	}
	result := 0
	record := [128]int{}
	start := 0

	for end := 0; end < len(s); end++ {
		// start index = max( current start index, previous current char index)
		start = max(start, record[s[end]])
		result = max(result, end-start+1) // calculate cur substring length and compare result
		record[s[end]] = end + 1          // record same char position + 1
	}

	return result
}

// Use slide window2:
// Time Complexity: O(n)
// Space Complexity:O(k)
// k for character set
// Runtime: 3 ms, faster than 90.23%
// Memory Usage: 2.7 MB, less than 100.00%
func lengthOfLongestSubstring2(s string) int {
	arr := make([]int, 128)
	result := 0

	l, r := 0, 0

	for r < len(s) {
		idx := s[r]
		l = max(l, arr[idx])
		result = max(result, r-l+1)

		arr[idx] = r + 1
		r++
	}

	return result
}
